Instrument jl_gc_big_alloc() when called from generated code for Allocations Profiler

NHDaly opened the pull request:
Instrument jl_gc_big_alloc() when called from generated code for Allocations Profiler

Add _maybe_record_alloc_to_profile() call into jl_gc_big_alloc() to
allow us to instrument it in the Allocations Profiler.

Followup to #43868

Finishes last case of #43688 (gets the stacks, though not the types, of big allocs).

--------------------------


## Before

![graphviz (2)](https://user-images.githubusercontent.com/1582097/152557258-4a9669d8-fedc-4b43-bf37-6426982b16b5.svg)

*: instrumented; red: currently missed; green: newly introduced function

<details>
<summary>graphviz</summary>

```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_gc_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc *"];
  jl_gc_pool_alloc_noinline [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc_noinline"];
  jl_gc_pool_alloc_inner;
  jl_gc_big_alloc;

  {
    rank=same;
    jl_gc_managed_malloc;
    jl_gc_pool_alloc_noinline;
    jl_gc_big_alloc;
  }

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc [color=red];
  generated_code -> jl_gc_pool_alloc;
  jl_gc_pool_alloc -> jl_gc_pool_alloc_inner  [style=dotted label="inlined"];
  jl_gc_pool_alloc_noinline -> jl_gc_pool_alloc_inner  [style=dotted  label="inlined"];


  jl_gc_alloc -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc -> jl_gc_big_alloc;
  new_array -> jl_gc_alloc;
  new_array -> jl_gc_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc_string -> jl_gc_big_alloc;
}
```
</details>

## After

![graphviz (3)](https://user-images.githubusercontent.com/1582097/152869280-27ddc7cd-0707-4cce-939e-30f8885a361b.svg)

*: instrumented; green: newly introduced function

<details>
<summary>graphviz</summary>

```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_gc_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc *"];
  jl_gc_pool_alloc_noinline [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc_noinline"];
  jl_gc_pool_alloc_inner;
  jl_gc_big_alloc [fillcolor="#90ed8f" style=filled label="jl_gc_big_alloc *"];
  jl_gc_big_alloc_noinline [fillcolor="#90ed8f" style=filled label="jl_gc_pool_alloc_noinline"];
  jl_gc_big_alloc_inner;


  {
    rank=same;
    jl_gc_managed_malloc;
    jl_gc_pool_alloc_noinline;
    jl_gc_big_alloc_noinline;
  }

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc;
  generated_code -> jl_gc_pool_alloc;
  jl_gc_pool_alloc -> jl_gc_pool_alloc_inner  [style=dotted label="inlined"];
  jl_gc_pool_alloc_noinline -> jl_gc_pool_alloc_inner  [style=dotted  label="inlined"];
  jl_gc_big_alloc -> jl_gc_big_alloc_inner  [style=dotted label="inlined"];
  jl_gc_big_alloc_noinline -> jl_gc_big_alloc_inner  [style=dotted  label="inlined"];


  jl_gc_alloc -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc -> jl_gc_big_alloc_noinline;
  new_array -> jl_gc_alloc;
  new_array -> jl_gc_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc_noinline;
  jl_gc_alloc_string -> jl_gc_big_alloc_noinline;
}
```
</details>

NHDaly committed:
Instrument jl_gc_big_alloc() when called from generated code for Allocations Profiler

Add _maybe_record_alloc_to_profile() call into jl_gc_big_alloc() to
allow us to instrument it in the Allocations Profiler.

Followup to #43868

Finishes fixing #43688 (gets the stacks, though not the types, of big allocs).

-----

- Rename jl_gc_big_alloc to jl_gc_big_alloc_inner, make it inlined
- add jl_gc_pool_alloc_noinline that calls jl_gc_big_alloc_inner
    - replace all internal calls to jl_gc_big_alloc to call this instead
- add a new jl_gc_pool_alloc that calls jl_gc_big_alloc_inner
    - This one is instrumented, and is meant to be called by generated
      code.

vilterp reviewed (approved):
this looks in line (pun somewhat intended) with what we do for pool alloc :)

NHDaly commented:
@nanosoldier `runbenchmarks("alloc", vs = ":pv-alloc-profile-wrap-pool-alloc")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44043#issuecomment-1030270421) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/adf2746_vs_c0c854e/report.md).

vilterp commented:
Kind of hard to believe this isn't just noise… Like big alloc gets called waaay less than pool alloc — how could instrumenting pool alloc not cause a regression, but this does? 🤔

NHDaly commented:
Yep, 100% agreed. The Nanosoldier report for the `pool_alloc` PR showed a `5%` improvement, which is almost certainly impossible, and this one shows a `5%` regression. So i think both are almost certainly lost in the noise 👍 👍

NHDaly commented on src/gc.c:
```diff
@@ -972,6 +972,22 @@ JL_DLLEXPORT jl_value_t *jl_gc_big_alloc(jl_ptls_t ptls, size_t sz)
+// This wrapper exists only to prevent `jl_gc_big_alloc_inner` from being inlined into
+// its callers. We provide an external-facing interface for callers, and inline `jl_gc_big_alloc_inner`
+// into this. (See https://github.com/JuliaLang/julia/pull/43868 for more details.)
+jl_value_t *jl_gc_big_alloc_noinline(jl_ptls_t ptls, size_t sz) {
```
Oh, oops, good call. I'm glad you've reviewed it! 😁 thanks. You're right, i missed this part when i copied the previous PR! lemme fix that.

NHDaly replied:
Oh, oops, good call. I'm glad you've reviewed it! 😁 thanks. You're right, i missed this part when i copied the approach from the previous PR!

Lemme fix that and ping you again when it's done. Thank you!

NHDaly replied:
Okay! This looks good to go now. Thanks for catching that.
(And thanks for the help, @vilterp)

vilterp committed:
fix use of pool_alloc_noinline where big_alloc_noinline was intended (#44066)

NHDaly commented:
K, I've also updated the diagrams at the top of the PR description. This should be ready for a re-review @vtjnash. Thanks again! 👍

vchuravy commented on src/julia_internal.h:
```diff
@@ -231,7 +231,7 @@ JL_DLLEXPORT extern const char *jl_filename;
 
 jl_value_t *jl_gc_pool_alloc_noinline(jl_ptls_t ptls, int pool_offset,
                                    int osize);
-JL_DLLEXPORT jl_value_t *jl_gc_big_alloc(jl_ptls_t ptls, size_t allocsz);
```
This seems wrong? You want to dllexport `jl_gc_big_alloc`

NHDaly replied:
Eep, i think you're mostly right! Gosh sorry this is so tricky. C declarations...

You're right: i want to dllexport `jl_gc_big_alloc`. But I _am_ already doing that, because the _definition_ is marked `dllexport`.

You're also right that I _shouldn't_ be dllexporting `jl_gc_big_alloc_noinline`. I'll remove that now.

But _I'm right_ (I'm pretty sure) that we want to delete jl_gc_big_alloc from the header file, since it's not meant to be called directly from the C code anywhere, so it doesn't need to be forward declared. It's simply emitted by the LLVM codegen, and since it's dllexported in the _object file_, it gets linked correctly! :)

So i think the only change to make here is to drop the `JL_DLLEXPORT` from jl_gc_big_alloc_noinline.

I'll do that now.

-----

This is the pattern we followed in the previous PR:
https://github.com/JuliaLang/julia/pull/43868/files#diff-7b3edb5362e7b70b75aff15c04b492aa3c74b883fca8eeee74d7bd839f81e4b0L231-R232

NHDaly replied:
I've made this change. thanks!

NHDaly commented on src/julia_internal.h:
```diff
@@ -231,7 +231,7 @@ JL_DLLEXPORT extern const char *jl_filename;
 jl_value_t *jl_gc_pool_alloc_noinline(jl_ptls_t ptls, int pool_offset,
                                    int osize);
-JL_DLLEXPORT jl_value_t *jl_gc_big_alloc(jl_ptls_t ptls, size_t allocsz);
+JL_DLLEXPORT jl_value_t *jl_gc_big_alloc_noinline(jl_ptls_t ptls, size_t allocsz);
```
Dropping JL_DLLEXPORT from jl_gc_big_alloc_noinline declaration, per @vchuravy's previous comment, above.
```suggestion
jl_value_t *jl_gc_big_alloc_noinline(jl_ptls_t ptls, size_t allocsz);
```

NHDaly committed:
Update src/julia_internal.h - PR review feedback

Dropping JL_DLLEXPORT from jl_gc_big_alloc_noinline declaration, per @vchuravy's previous comment, above.

NHDaly commented:
Thanks everyone! Merging now 👍

vchuravy commented on src/julia_internal.h:
```diff
@@ -231,7 +231,7 @@ JL_DLLEXPORT extern const char *jl_filename;
 
 jl_value_t *jl_gc_pool_alloc_noinline(jl_ptls_t ptls, int pool_offset,
                                    int osize);
-JL_DLLEXPORT jl_value_t *jl_gc_big_alloc(jl_ptls_t ptls, size_t allocsz);
```
I think you technically want to keep this line still? `jl_gc_big_alloc` is still an exported function?

NHDaly replied:
🤔 it is, but it's not meant to be called anywhere from inside our C code, only from LLVM-generated code. So it doesn't need to be declared in `julia_internal.h`, which, from what i understand, is a private header only for internal functions?

It's still exported via the definition, which is all that really matters to the linker anyway, this declaration is only here for compiling C files.

So i think it's okay to delete, like we did. We did the same thing for `pool_alloc` in [#43688](https://github.com/JuliaLang/julia/pull/43868), too.
